### PR TITLE
[chore] [exporter/splunkhec] fix flaky test expecting all requests to be gzipped.

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Fix flaky test expecting all requests to be gzipped. The last request might not be zipped.

**Link to tracking Issue:**
#17177

**Testing:**
Ran the test locally and didn't see a flaky behavior anymore.

**Documentation:**
N/A

### CODE REVIEW

File: exporter/splunkhecexporter/exporter_test.go [hunk 1]
```diff
@@ -201,23 +202,28 @@ func TestConsumeMetricsData(t *testing.T) {
...
-				zipReader, err := gzip.NewReader(bytes.NewReader(body))
-				assert.NoError(t, err)
-				bodyBytes, _ := io.ReadAll(zipReader)
-				firstPayload := strings.Split(string(bodyBytes), "}{")[0]
-				var metric splunk.Event
-				err = json.Unmarshal([]byte(firstPayload+"}"), &metric)
-				if err != nil {
-					t.Fatal(err)
+				bodyBytes := body
+				// the last batch might not be zipped.
```
Comment (atoulme) [inline exporter/splunkhecexporter/exporter_test.go:211]: this created flakiness. Sometimes the last request sent by the batch was not gzipped. We account for it.

File: exporter/splunkhecexporter/exporter_test.go [hunk 2]
```diff
@@ -201,23 +202,28 @@ func TestConsumeMetricsData(t *testing.T) {
...
+				if "gzip" == r.Header.Get("Content-Encoding") {
+					zipReader, err := gzip.NewReader(bytes.NewReader(body))
+					require.NoError(t, err)
+					bodyBytes, _ = io.ReadAll(zipReader)
+				}
+
+				events := strings.Split(string(bodyBytes), "}{")
+				firstPayload := events[0]
```
Comment (atoulme) [inline exporter/splunkhecexporter/exporter_test.go:219]: also sometimes the last batch had just one event. We fix it here by doing a better job separating HEC events.

File: exporter/splunkhecexporter/client.go [hunk 3]
```diff
@@ -78,6 +80,10 @@ func (b *bufferState) Read(p []byte) (n int, err error) {
 }
 
 func (b *bufferState) Close() error {
+	if b.closed {
+		return nil
+	}
+	b.closed = true
```
Comment (dmitryax) [inline exporter/splunkhecexporter/client.go:86]: Is this change really needed? Doesn't `gzip.Writer` already do that?

Comment (dmitryax) [reply exporter/splunkhecexporter/client.go:86]: I'm looking at https://github.com/golang/go/blob/master/src/compress/gzip/gzip.go#L228

**Comment by dmitryax** [comment]
Please check linter failures

File: exporter/splunkhecexporter/client.go (hunk 3 above)
Comment (atoulme) [reply exporter/splunkhecexporter/client.go:86]: oh you got it indeed.

**Review by dmitryax — APPROVED** [review]
